build/kconfig/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// Application configuration passed into the kernel build.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct KernelConfig {
+    /// Tasks in the app image. The order of tasks is significant.
+    pub tasks: Vec<TaskConfig>,
+
+    /// Regions that tasks have shared access to, keyed by the name the task
+    /// config used to grant access (often peripheral name). These are typically
+    /// memory mapped peripherals (and should probably not be anything else).
+    pub shared_regions: BTreeMap<String, RegionConfig>,
+
+    /// Interrupts hooked by the application, keyed by IRQ number.
+    pub irqs: BTreeMap<u32, InterruptConfig>,
+}
+
+/// Configuration for a single hooked interrupt.
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    Serialize,
+    Deserialize,
+    Eq,
+    PartialEq,
+    Hash,
+    Ord,
+    PartialOrd,
+)]
+pub struct InterruptConfig {
+    /// Index of task (in the application task array) that receives this
+    /// interrupt.
+    pub task_index: usize,
+    /// Notification bits that are posted to the task when the interrupt fires.
+    /// Note that this is a mask and can have multiple (or zero!) bits set; the
+    /// kernel doesn't really care.
+    pub notification: u32,
+}
+
+/// Record describing a single task.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct TaskConfig {
+    /// Named memory regions that this task has exclusive access to, keyed by
+    /// name.
+    ///
+    /// The name is the "output" assignment that generated this region,
+    /// typically (but not necessarily!) either `"ram"` or `"flash"`.
+    pub owned_regions: BTreeMap<String, RegionConfig>,
+
+    /// Names of regions (in the app-level `shared_regions`) that this task
+    /// needs access to.
+    pub shared_regions: BTreeSet<String>,
+
+    /// Address of the task's entry point. This is the first instruction that
+    /// will be executed whenever the task is (re)started.
+    pub entry_point: OwnedAddress,
+
+    /// Address of the task's initial stack pointer, to be loaded at (re)start.
+    /// It must be pointing into or *just past* one of the task's memory
+    /// regions.
+    pub initial_stack: OwnedAddress,
+
+    /// Initial priority of this task.
```

Comment:
"Initial" suggests priority may change?

Reply:
Yeah, doing priority ceiling protocol may require us to change the dynamic priority of tasks, if we decide to do that.